src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs

```diff
@@ -645,26 +645,29 @@ private Uri GetContainingEntitySetUri(Uri baseUri, ODataUri odataUri)
         {
             ODataPath path = odataUri.Path;
             List<ODataPathSegment> segments = path.ToList();
+            int segmentsCount = segments.Count - 1;
             ODataPathSegment lastSegment = segments.Last();
             while (!(lastSegment is NavigationPropertySegment) && !(lastSegment is OperationSegment))
             {
-                segments.Remove(lastSegment);
-                lastSegment = segments.Last();
+                segmentsCount--;
+                lastSegment = segments[segmentsCount];
             }
 
             // also removed the last navigation property segment
-            segments.Remove(lastSegment);
+            segments = segments.GetRange(0, segmentsCount);

```

Comment:
@marabooy I wish there was a way the change starting from [this](https://github.com/OData/odata.net/pull/1854/files#diff-5c1c8c475ffa3bfbcec9695a7554881aR648) line to here could be isolated and unit tested. This is the stuff off-by-one bugs are made off - if you got your arithmetic wrong!

Reply:
I added a test to verify that the url is not impacted, earlier while working the wrong arithmetic could be observed on nested navigation properties where some of the segments repeated themselved. The test case will catch this if it happens again.